admin/server/github.go

```diff
@@ -413,12 +423,96 @@ func (s *Server) CreateManagedGitRepo(ctx context.Context, req *adminv1.CreateMa
 	}
 
 	return &adminv1.CreateManagedGitRepoResponse{
-		Remote:   *repo.CloneURL,
-		Username: "x-access-token",
-		Password: token,
+		Remote:        *repo.CloneURL,
+		Username:      "x-access-token",
+		Password:      token,
+		DefaultBranch: valOrDefault(repo.DefaultBranch, "main"),
 	}, nil
 }
 
+// DisconnectProjectFromGithubRequest disconnects a project from Github by uploading the contents of a Github repository to a rill managed repository.
+func (s *Server) DisconnectProjectFromGithub(ctx context.Context, req *adminv1.DisconnectProjectFromGithubRequest) (*adminv1.DisconnectProjectFromGithubResponse, error) {
+	observability.AddRequestAttributes(ctx,
+		attribute.String("args.organization", req.Organization),
+		attribute.String("args.project", req.Project),
+	)
+
+	// Check the request is made by a user
+	claims := auth.GetClaims(ctx)
+	if claims.OwnerType() != auth.OwnerTypeUser {
+		return nil, status.Error(codes.Unauthenticated, "not authenticated as a user")
+	}
+
+	// Find parent org
+	proj, err := s.admin.DB.FindProjectByName(ctx, req.Organization, req.Project)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	// Check permissions
+	if !claims.ProjectPermissions(ctx, proj.OrganizationID, proj.ID).ManageProject {
+		return nil, status.Error(codes.PermissionDenied, "does not have permission to edit project")
+	}
+
+	if proj.GithubURL == nil || proj.ManagedGitRepoID != nil {
+		return nil, status.Error(codes.InvalidArgument, "project is not connected to github")
+	}
+
+	// create a managed git repo
+	org, err := s.admin.DB.FindOrganization(ctx, proj.OrganizationID)
+	if err != nil {
+		return nil, err
+	}
+
+	repo, err := s.admin.CreateManagedGitRepo(ctx, org, proj.Name, claims.OwnerID())
+	if err != nil {
+		return nil, err
+	}
+	id, err := s.admin.Github.ManagedOrgInstallationID()
+	if err != nil {
+		return nil, err
+	}
+
+	mgdRepoToken, err := s.admin.Github.InstallationToken(ctx, id, *repo.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	copyData := func(path string) error {
+		// download and copy to a tempo location
+		repoID, err := s.githubRepoIDForProject(ctx, proj)
+		if err != nil {
+			return err
+		}
+		token, err := s.admin.Github.InstallationToken(ctx, *proj.GithubInstallationID, repoID)
+		if err != nil {
+			return err
+		}
+
+		return copyFromSrcGit(path, *proj.GithubURL, proj.ProdBranch, proj.Subpath, token)
+	}
+	err = s.pushToGit(ctx, copyData, *repo.CloneURL, *repo.DefaultBranch, "", mgdRepoToken, true)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
I'm wondering if this logic could be decoupled/simplified some more. Perhaps moved to a Git utils package.

For example, could we separate the clone and re-upload logic into separate utils? Like have `gitutils.ShallowClone` and `gitutils.ForcePush`, which we connect together here, or something like that?